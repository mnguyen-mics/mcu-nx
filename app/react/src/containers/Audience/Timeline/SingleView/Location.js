import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Row, Col } from 'antd';
import ReactMapGL, { Marker } from 'react-map-gl';
import Dimensions from 'react-dimensions';

class LocationHelper extends Component {

  render() {
    const {
    longitude,
    latitude,
  } = this.props;

    const settings = {
      dragPan: false,
      dragRotate: false,
      scrollZoom: false,
      touchZoomRotate: false,
      doubleClickZoom: false,
    };

    return (longitude && latitude) ? (
      <Row gutter={10} className="table-line section border-top">
        <Col span={24} style={{ height: '100px', margin: '-5px -20px' }}>
          <ReactMapGL
            {...settings}
            style={{ cursor: 'default' }}
            width={this.props.containerWidth + 40}
            height={100}
            latitude={latitude}
            longitude={longitude}
            zoom={9}
            interactive={false}
            attributionControl={false}
            preventStyleDiffing
            mapboxApiAccessToken={global.window.MCS_CONSTANTS.MAPBOX_TOKEN}
          >
            <Marker
              latitude={latitude}
              longitude={longitude}
            >
              <svg width="10" height="10">
                <circle cx="5" cy="5" r="5" fill="#00a1df" />
              </svg>
            </Marker>
          </ReactMapGL>
        </Col>
      </Row>
  ) : (<span />);
  }
}

LocationHelper.defaultProps = {
  location: null,
  longitude: 0,
  latitude: 0,
};

LocationHelper.propTypes = {
  longitude: PropTypes.number,
  latitude: PropTypes.number,
  containerWidth: PropTypes.number.isRequired,
};

export default Dimensions()(LocationHelper);
