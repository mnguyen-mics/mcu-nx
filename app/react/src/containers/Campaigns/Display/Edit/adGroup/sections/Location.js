import React from 'react';
import { Row } from 'antd';

import FormSection from '../../../../../../components/Partials/FormSection';
import messages from '../../messages';

function Location() {

  return (
    <div id="location">
      <FormSection
        subtitle={messages.sectionSubtitle4}
        title={messages.sectionTitle4}
      />
      <Row />
    </div>
  );
}

export default Location;
