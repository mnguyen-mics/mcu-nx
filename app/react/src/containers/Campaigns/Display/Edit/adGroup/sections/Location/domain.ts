import { Omit } from './../../../../../../../utils/Types';
export interface LocationSelectionResource {
  id: string;
  geoname_id: string;
  country: string;
  admin1: string;
  admin2: string;
  exclude: boolean;
}

export interface LocationFieldModel {
  id: string;
  resource: LocationSelectionResource | Omit<LocationSelectionResource, 'id'>;
  deleted: boolean;
}
