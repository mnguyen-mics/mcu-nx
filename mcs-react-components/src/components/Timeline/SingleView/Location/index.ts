import LocationHelper from './Location';

export { LocationProps } from './Location';

export default LocationHelper;
