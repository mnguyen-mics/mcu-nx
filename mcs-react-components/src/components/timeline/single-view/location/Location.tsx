import * as React from 'react';
import { Row, Col } from 'antd';
import ReactMapGL, { Marker } from 'react-map-gl';

export interface LocationProps {
  longitude: number;
  latitude: number;
  containerWidth?: number;
  mapboxToken?: string;
}

class Location extends React.Component<LocationProps> {
  render() {
    const { longitude, latitude, mapboxToken } = this.props;

    const settings = {
      dragPan: false,
      dragRotate: false,
      scrollZoom: false,
      touchZoomRotate: false,
      doubleClickZoom: false,
    };

    const containerWidth = this.props.containerWidth ? this.props.containerWidth : 0;

    const mapToken =
      ((global as any).window.MCS_CONSTANTS && (global as any).window.MCS_CONSTANTS.MAPBOX_TOKEN) ||
      mapboxToken;

    return longitude && latitude && mapToken ? (
      <Row gutter={10} className='table-line section border-top'>
        <Col span={24} style={{ height: '100px', margin: '-5px -20px' }}>
          <ReactMapGL
            {...settings}
            width={containerWidth + 40}
            height={100}
            latitude={latitude}
            longitude={longitude}
            zoom={9}
            attributionControl={false}
            preventStyleDiffing={true}
            mapboxApiAccessToken={mapToken}
          >
            <Marker latitude={latitude} longitude={longitude}>
              <svg width='10' height='10'>
                <circle cx='5' cy='5' r='5' fill='#00a1df' />
              </svg>
            </Marker>
          </ReactMapGL>
        </Col>
      </Row>
    ) : (
      <span />
    );
  }
}

export default Location;
