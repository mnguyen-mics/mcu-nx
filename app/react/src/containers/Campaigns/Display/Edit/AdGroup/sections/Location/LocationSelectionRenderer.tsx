import * as React from 'react';
import { Row } from 'antd';
import { ButtonStyleless } from '../../../../../../../components';
import McsIcons, { McsIconType } from '../../../../../../../components/McsIcons';
import { LocationFieldModel } from '../../domain';
import { Geoname } from '../../../../../../../services/GeonameService';
import GeonameRenderer from '../../../../../../../containers/Geoname/GeonameRenderer';

interface Props {
  locationFields: LocationFieldModel[];
  onClickOnRemove: (locationField: LocationFieldModel, index: number) => void;
}

class LocationSelectionRenderer extends React.Component<Props> {

  static defaultProps: Partial<Props> = {
    locationFields: [],
  };

  render() {

    const {
      locationFields,
      onClickOnRemove,
    } = this.props;

    return (
      <div>
        <Row
          type="flex"
          align="middle"
          justify="space-between"
          className={locationFields.length !== 0 ? 'search-result-box' : 'hide-section'}
        >
          {locationFields.map((locationField, index) => {

            const handleOnClick = () => onClickOnRemove(locationField, index);
            const iconType: McsIconType = locationField.model.exclude ? 'close-big' : 'check';
            const renderGeoname = (geoname: Geoname) => <span>{geoname.name}</span>;

            return (
              <div className={'search-result-box-item'} key={locationField.key}>
                <McsIcons type={iconType} />
                <GeonameRenderer
                  key={locationField.key}
                  geonameId={locationField.model.geoname_id}
                  renderMethod={renderGeoname}
                />
                <ButtonStyleless
                  className="close-button"
                  onClick={handleOnClick}
                >
                  <McsIcons type="close" />
                </ButtonStyleless>
              </div>
            );
          })}
        </Row>
      </div>
    );
  }
}

export default LocationSelectionRenderer;
