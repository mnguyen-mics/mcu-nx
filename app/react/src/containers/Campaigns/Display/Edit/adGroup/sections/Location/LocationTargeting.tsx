import * as React from 'react';
import {  Row, Col, Checkbox, Tooltip, Modal } from 'antd';
import { WrappedFieldArrayProps, InjectedFormProps } from 'redux-form';
import { FormattedMessage, injectIntl, InjectedIntlProps } from 'react-intl';

import { McsIcons } from '../../../../../../../components';
import messages from '../../../messages';
import LocationSelectionRenderer from './LocationSelectionRenderer';
import { isFakeId } from '../../../../../../../utils/FakeIdHelper';
import GeonameRenderer from '../../../../../../Geoname/GeonameRenderer';
import SelectGeoname from './SelectGeoname';
import { LocationFieldModel } from './domain';
import { Geoname } from '../../../../../../../services/GeonameService';

const confirm = Modal.confirm;

export interface LocationTargetingProps {
  RxF: InjectedFormProps;
}

interface LocationTargetingState {
  locationTargetingDisplayed: boolean;
}

type JoinedProps =
  LocationTargetingProps &
  InjectedIntlProps &
  WrappedFieldArrayProps<LocationFieldModel>;

class LocationTargeting extends React.Component<JoinedProps, LocationTargetingState> {

  constructor(props: JoinedProps) {
    super(props);
    this.state = {
      locationTargetingDisplayed: false,
    };
  }

  markAsDeleted = (locationField: LocationFieldModel) => {
    const {
      fields,
      RxF,
    } = this.props;

    const locationFields = fields.getAll();
    const newLocationFields: LocationFieldModel[] = [];

    locationFields.forEach(field => {
      const found = field.id === locationField.id;
      if (found && !isFakeId(field.id) && field.deleted !== true) {
        newLocationFields.push({
          ...field,
          deleted: true,
        });
      } else if (!found) {
        newLocationFields.push(field);
      }
    });
    RxF.change((fields as any).name, newLocationFields);
  }

  addLocationField = (locationField: LocationFieldModel) => {
    const { fields, intl: { formatMessage } } = this.props;
    const allFields = fields ? fields.getAll().filter(item => !item.deleted) : [];
    const allFieldsWithDeleted = fields ? fields.getAll() : [];

    const formattedLocationField = {
      ...locationField,
    };

    let isUnderneath = false;
    let parentIsExcluded = false;
    const idsToRemove: string[] = [];
    allFields.forEach(item => {
      if (item.resource.country === locationField.resource.country) {
        if (item.resource.admin1 === '00') {
          // is coutry
          isUnderneath = true;
          if (item.resource.exclude) {
            parentIsExcluded = true;
          }
        } else if (item.resource.admin1 === locationField.resource.admin1 && locationField.resource.admin2 && !item.resource.admin2) {
          // is same admin1
          isUnderneath = true;
          if (item.resource.exclude) {
            parentIsExcluded = true;
          }
        }
      }
    });

    // need to exclude
    if (!parentIsExcluded && isUnderneath) {
      formattedLocationField.resource.exclude = true;
    }

    allFields.forEach(item => {
      if (item.resource.country === locationField.resource.country) {
        if (locationField.resource.admin1 === '00') {
          // adding a country so we need to remove subsequent underneath locations
          if (locationField.resource.exclude) {
            idsToRemove.push(item.id);
          } else if (!item.resource.exclude) {
            idsToRemove.push(item.id);
          }

        } else if (locationField.resource.admin1 === item.resource.admin1 && locationField.resource.admin2 === null) {
          // is admin 1 and has subsequent underneath location
          if (locationField.resource.exclude) {
            idsToRemove.push(item.id);
          } else if (!item.resource.exclude) {
            idsToRemove.push(item.id);
          }
        }
      }
    });

    if (parentIsExcluded) {
      Modal.warning({
        title: formatMessage(messages.contentSectionLocationModal1Title),
        content: formatMessage(messages.contentSectionLocationModal1),
      });
    } else if (idsToRemove.length) {
      const geonameIds = allFields.reduce((acc: string[], val, i) => {
        if (idsToRemove.includes(val.id)) {
          return [...acc, val.resource.geoname_id];
        }
        return acc;
      }, []);
      const renderGeoname = (el: Geoname) => (<div>{el.name}</div>);
      const content = (
        <div>
          <div>{formatMessage(messages.contentSectionLocationModal2)}</div>
          <br />
          {
            geonameIds.map(id => {
              return (<GeonameRenderer key={id} geonameId={id} renderMethod={renderGeoname} />);
            })
          }

        </div>
      );
      Modal.confirm({
        title: formatMessage(messages.contentSectionLocationModal2Title),
        content: content,
        onOk: () => {
          const newFields = allFieldsWithDeleted.filter(item => !idsToRemove.includes(item.id));
          this.props.RxF.change((fields as any).name, newFields.concat([formattedLocationField]));
        },
      });
    } else {
      this.props.RxF.change((fields as any).name, allFieldsWithDeleted.concat([formattedLocationField]));
    }

  }

  toggleDisplayLocationTargetingSection = () => {
    this.setState({
      locationTargetingDisplayed: !this.state.locationTargetingDisplayed,
    });
  }

  getDisplayedLocations = (locationFields: LocationFieldModel[]) => {
    return locationFields.filter(f => !f.deleted);
  }

  handleCheckbox = () => {
    const {
      intl: {
        formatMessage,
       },
      fields,
      RxF,
    } = this.props;

    if (this.getDisplayedLocations(fields.getAll()).length > 0) {
      confirm({
        cancelText: formatMessage(messages.cancel),
        content: formatMessage(messages.notificationWarning),
        maskClosable: true,
        okText: formatMessage(messages.ok),
        onOk: () => {
          const locationFields = fields.getAll();
          const newLocationFields: LocationFieldModel[] = [];

          locationFields.forEach(field => {
            if (!isFakeId(field.id)) {
              newLocationFields.push({
                ...field,
                deleted: true,
              });
            }
          });
          RxF.change((fields as any).name, newLocationFields);
        },
      });
    } else if (fields.length === 0) {
      this.setState({
        locationTargetingDisplayed: !this.state.locationTargetingDisplayed,
      });
    } else {
      this.toggleDisplayLocationTargetingSection();
    }
  }

  render() {

    const {
      fields,
      intl: {
        formatMessage,
      },
    } = this.props;

    const {
      locationTargetingDisplayed,
    } = this.state;

    const locationFields = fields.getAll();
    const showLocationTargeting =
      (locationTargetingDisplayed || this.getDisplayedLocations(locationFields).length > 0);

    const excludedGeonamesIds = this.getDisplayedLocations(locationFields).map(field => field.resource.geoname_id);

    return (
      <div>
        <Checkbox
          checked={showLocationTargeting}
          className="field-label checkbox-location-section"
          onChange={this.handleCheckbox}
        >
          <FormattedMessage id="location-checkbox-message" defaultMessage="I want to target a specific location" />
        </Checkbox>
        <div className={showLocationTargeting ? '' : 'hide-section'}>
          <Row align="middle" type="flex">
            <Col span={4} />
            <Col span={10} >
              <LocationSelectionRenderer
                locationFields={locationFields}
                onClickOnRemove={this.markAsDeleted}
              />
            </Col>
          </Row>
          <Row align="middle" type="flex">
            <Col span={4} className="label-col field-label">
              <FormattedMessage id="label-location-targeting" defaultMessage="Location : " />
            </Col>
            <Col span={10}>
              <SelectGeoname
                onGeonameSelect={this.addLocationField}
                excludedGeonamesIds={excludedGeonamesIds}
              />
            </Col>
            <Col span={2} className="field-tooltip">
              <Tooltip placement="right" title={formatMessage(messages.contentSectionLocationTooltipMessage)}>
                <McsIcons type="info" />
              </Tooltip>
            </Col>
          </Row>
        </div>
      </div>
    );
  }
}

export default injectIntl(LocationTargeting);
