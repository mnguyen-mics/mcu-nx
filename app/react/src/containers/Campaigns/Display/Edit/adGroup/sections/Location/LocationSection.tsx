import * as React from 'react';

import { GenericFieldArray, Field, FieldArray, InjectedFormProps } from 'redux-form';

import messages from '../../../messages';
import LocationTargeting, { LocationTargetingProps } from './LocationTargeting';
import { FormSection } from '../../../../../../../components/Form';

const LocationTargetingFieldArray = FieldArray as new() => GenericFieldArray<Field, LocationTargetingProps>;

export interface Props {
  RxF: InjectedFormProps;
}

class LocationSection extends React.Component<Props> {

  render() {

    return (
      <div id="locationTargeting" className="locationTargeting">
        <FormSection
          subtitle={messages.sectionSubtitleLocation}
          title={messages.sectionTitleLocationTargeting}
        />

        <LocationTargetingFieldArray
          name="locationTargetingTable"
          component={LocationTargeting}
          RxF={this.props.RxF}
        />

      </div>

    );
  }
}

export default LocationSection;
