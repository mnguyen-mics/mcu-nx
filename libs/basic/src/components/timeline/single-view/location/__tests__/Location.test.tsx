import 'jest';
import React from 'react';
import TestRenderer from 'react-test-renderer';
import Location, { LocationProps } from '../Location';

it('renders the Location', () => {
  const props: LocationProps = {
    longitude: -122.45,
    latitude: 37.78,
    containerWidth: 800,
    mapboxToken:
      'pk.eyJ1Ijoiem9iaW9uZTE0MyIsImEiOiJjajZtMHZpYngxcm4yMndvMXVibXJwZGx3In0.9s_vDS6h2pU5ampVcydvCA',
  };
  const component = TestRenderer.create(<Location {...props} />);
  const res = component.toJSON();
  expect(res).toMatchSnapshot();
});
