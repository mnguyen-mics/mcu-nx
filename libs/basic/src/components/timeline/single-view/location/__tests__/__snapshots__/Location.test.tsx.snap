// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Location 1`] = `
<div
  className="ant-row table-line section border-top "
  style={
    Object {
      "marginLeft": -5,
      "marginRight": -5,
    }
  }
>
  <div
    className="ant-col ant-col-24"
    style={
      Object {
        "height": "100px",
        "margin": "-5px -20px",
        "paddingLeft": 5,
        "paddingRight": 5,
      }
    }
  >
    <div
      style={
        Object {
          "cursor": "grab",
          "height": 100,
          "position": "relative",
          "width": 840,
        }
      }
    >
      <div
        style={
          Object {
            "height": 100,
            "position": "relative",
            "width": 840,
          }
        }
      >
        <div
          style={
            Object {
              "height": 100,
              "visibility": "visible",
              "width": 840,
            }
          }
        />
        <div
          className="overlays"
          style={
            Object {
              "height": 100,
              "left": 0,
              "overflow": "hidden",
              "position": "absolute",
              "top": 0,
              "width": 840,
            }
          }
        >
          <div
            className="mapboxgl-marker "
            style={
              Object {
                "left": 419.9999999999951,
                "position": "absolute",
                "top": 50.0000000000194,
              }
            }
          >
            <svg
              height="10"
              width="10"
            >
              <circle
                cx="5"
                cy="5"
                fill="#00a1df"
                r="5"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
